### PR TITLE
Added impementation for the Criteria Query in Spring Boot task.

### PR DESCRIPTION
**Pull request by MinoDentori**
1. Added Search endpoint to the BookController class.
2. Added service layer implementation for the new endpoint.
3. Created SpecificationBulder and implementation class.
4. Created SpecificatoinProvider interface and implementations: AuthorSpecificationProvider, TitleSpecificationProvider, IsbnSpecificationProvider for filtration by corresponding Book field.
5. Created SpecificationProviderManager for better control of getting specific SpecificationProvider implementation.
6. Tested via Postman.

### CODE REVIEW

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import mate.academy.onlinebookstore.model.Book;
+import mate.academy.onlinebookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: Please make `"author"` as constant

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("author").in(Arrays.stream(params).toArray());
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: Please make `"author"` as constant

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import mate.academy.onlinebookstore.model.Book;
+import mate.academy.onlinebookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class IsbnSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "isbn";
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java]: Please make `"isbn"` as constant

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("isbn").in(Arrays.stream(params).toArray());
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java]: Please make `"isbn"` as constant

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import mate.academy.onlinebookstore.model.Book;
+import mate.academy.onlinebookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: Please make `"title"` as constant

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("title").in(Arrays.stream(params).toArray());
```
Comment (d1sam) [inline src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: Please make `"title"` as constant

**Review by d1sam — CHANGES_REQUESTED** [review]
Well done, but please avoid String literals

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java (hunk 1 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: fixed

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java (hunk 3 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java]: fixed

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java (hunk 2 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: fixed

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java (hunk 5 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: fixed

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java (hunk 4 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java]: fixed

File: src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java (hunk 6 above)
Comment (MinoDentori) [reply src/main/java/mate/academy/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: fixed

**Review by kshuryhin — APPROVED** [review]
Looks good!

**Pull request merged**